# Arabidopsis-6mA-CpG-Detection-and-Comparison-Pipeline

### A note to the grader
This is a pipeline to reproduce our project. We know that you are aware of our situation. Our end goal for this project is to have a next flow in which you input a .bam file. Our steward/sponsor has not permitted us to share any of the data files outside of our Capstone group. For that reason we have included the nextflow file and a guide to what the pipeline does but without any data (as approved by Dr. Payne).

Inputs:
  These are the files we would provide you with
  ```
  1000.bam (~12.3 GB), 1000.bam.bai (~774 KB), arabidopsis.fasta (~115 MB), arabidopsis.fasta.fai (~195 bytes)
  ```

### Step 1: Align 
Pbmm2 is a tool created by PacBio to do genome alignment. Using the align command provide a path to the reference genome and a path to the unaligned .bam file generated by the sequencing machine. The output will be an aligned .bam file and a .bam.bai file (an indexed bam file used for further analysis).

  ```
  pbmm2 align [genome file] [bam file] [output filename] --sort --preset CCS
  Ex.
  $ pbmm2 align arabidopsis.fasta 1000.bam 1000_mapped.bam --sort --preset CCS
  Output: 1000_mapped.bam, 1000_mapped.bam.bai
  ```
Note, you do not need to input the .bam.bai files, as they are automatically incorporated by the program.

### Step 2: Get CpG output
Pb-cpg-tools:
Pb-cpg-tools is a tool created by PacBio to create CpG (5mC) methylation probabilities using the .bam files created from pbmm2. Using the aligned_bam_to_cpg_scores command and a path the aligned .bam file (the .bai file must also be in the same directory) it will create multiple output files. The files we will use for further analysis and visualization are the .bw (bigwig) file and .bed file. 

### Step 3: Get 6mA output
convert.py:
convert.py is a simple python script written by our team to modify the .bed file from the previous into the format needed for further analysis. The new format of the file is a simplified .bed file containing only the necessary information.



### Step 4: Generate a Matrix
computeMatrix is a command included in deeptools, a command line tool available through bioconda. This is an intermediate step to compute a matrix that is used to generate a heatmap. The command requires the .bw (bigwig) file created by pb-cpg-tools and the new .bed file that was created by using convert.py. The additional arguments ensure that the genomic regions of interest are centered on the heatmap and the amount of basepairs on each side of the center point. The command will use the files and arguments to create a .MAT (matrix) file that will be used in the next step.

### Step 5: Plot Heatmap
plotHeatmap:
plotHeatmap is a command included in deeptools, a command line tool available through bioconda. This command uses the .MAT file from the previous step and creates a heatmap image file (png, svg, or pdf) based on the colors specified in the command.


### Step 6: View Data with Integrative Genomics Viewer (IGV)


